README.MD

```diff
@@ -1,105 +1,252 @@
-  },
-  cogateway: {
-    "deployer": "0x...", /* Address of key that will deploy the contract */
-    "bounty": '1000000000000000000'
-  }
+const redeemer = '0x0000000000000000000000000redeemeraddress';
+const redeemAmount = '1000000000000';
+const beneficiary = '0x000000000000000beneficiaryaccountaddress';
+const gasPrice = '4';
+const gasLimit = '150000000';
+const hashLock = '00000000000000000000000000000000000000000000000000000000hashlock';
+const txOptionRedeem = {
+  from: redeemer,
+  value: '100', // This must be equal to the bounty amount!
 };
+
+facilitator
+  .redeem(
+    redeemer,
+    redeemAmount,
+    beneficiary,
+    gasPrice,
+    gasLimit,
+    hashLock,
+    txOptionRedeem,
+  )
+  .then(console.log)
+  .catch(console.log);
 ```
-### Executing setup. This may some time.
+
+Progressing the redemption after the state root has been transferred:
 ```js
-helper.setup(simpleToken, originConfig, auxiliaryConfig)
-.then(function (output) {
-  console.log("output", output);
-})
+const unlockSecret = '0x0000000000000000000000000000000000000000000000000000unlocksecret';
+const txOptionOriginRedeem = { from: '0x000000000facilitatoraddressonoriginchain' };
+const txOptionAuxiliaryRedeem = { from: '0x000000facilitatoraddressonauxiliarychain' };
+
+facilitator
+  .progressRedeem(
+    redeemer,
+    nonce,
+    beneficiary,
+    redeemAmount,
+    gasPrice,
+    gasLimit,
+    hashLock,
+    unlockSecret,
+    txOptionOriginRedeem,
+    txOptionAuxiliaryRedeem,
+  )
+  .then(console.log)
+  .catch(console.log);
 ```
 
+#### General Contract Interaction
+
+EIP20Gateway example:
+```js
+// Expecting Mosaic import and an instance of `mosaic`:
+const {
+  Anchor,
+  EIP20CoGateway,
+  EIP20Gateway,
+  EIP20Token,
+  OSTPrime,
+} = Mosaic.Contracts;
+
+// Afterwards, ou can call methods directly on the contract instance:
+const eip20Gateway = new EIP20Gateway(
+  mosaic.origin.web3,
+  mosaic.origin.contractAddresses.EIP20Gateway,
+);
+
+const staker = '0x00000000000000000000stakeraccountaddress';
+
+const amount = '1000000000000';
+const beneficiary = '0x000000000000000beneficiaryaccountaddress';
+const gasPrice = '5';
+const gasLimit = '150000000';
+const hashLock = '0x00000000000000000000000000000000000000000000000000000000hashlock';
+const txOptions = { from: staker };
+
+const nonce = await eip20Gateway.getNonce(staker);
+
+eip20Gateway
+  .stake(
+    amount,
+    beneficiary,
+    gasPrice,
+    gasLimit,
+    nonce,
+    hashLock,
+    txOptions,
+  )
+  .then(console.log)
+  .catch(console.log);
+```
+
+#### OSTPrime
+
+On the auxiliary chain OST is the base token.
+The gas price on auxiliary is paid in OST (like Ether pays for gas on Ethereum mainnet).
+To obtain the base token, you need to stake OST on origin.
+<!-- Add contract addresses for test nets -->
+On auxiliary, the `OSTPrime` contract will convert the EIP20 token to base tokens for the beneficiary.
```

Comment:
Should we provide a code snippet to show `wrap` and `unwrap`?

Reply:
If yes, I can help to provide the code snippet.

Reply:
Please do. Thank you! 🎉

Reply:
Create OSTPrime object
```
const ostPrime = new Mosaic.ContractInteract.OSTPrime(
  mosaic.auxiliary.web3,
  mosaic.auxiliary.contractAddresses.OSTPrime,
);

```
Wrap
```
const amountToWrap = '1000';
const fromAddress = '0x00000000000000000000000000000fromaddress';
let txOptions = {
  value: amountToWrap,
  from: fromAddress,
};

ostPrime.wrap(txOptionsAuxiliaryWithValue).then(console.log).catch(console.log);
```

Unwrap
```
const amountToUnwrap = '1000';
txOptions = {
  from: fromAddress,
};

ostPrime.unwrap(amountToUnwrap, txOptions).then(console.log).catch(console.log);
```